pkg/trace/writer/sender.go

```diff
@@ -247,6 +250,14 @@ func (s *sender) sendPayload(p *payload) {
 			return
 		}
 		atomic.AddInt32(&s.attempt, 1)
+		atomic.AddInt32(&p.retries, 1)
+
+		if atomic.LoadInt32(&p.retries) > maxRetries {
+			//exceeded max retries per payload, dropping payload
+			log.Error("Exceeded max retries per payload, dropping payload ")
+			s.releasePayload(p, eventTypeDropped, stats)
+			return
+		}
```

Comment:
We should consider how long it takes to reach `maxRetries`. 
It looks like the payload goes right back into the queue when it fails. If the backend is offline even for a minute, we will end up dropping payloads. We should probably hang onto them as long as we can and just let them fall out of the queue when it fills up.

Changing the `Debugf` to a `Warnf` in `pkg/trace/writer/trace.go` solves the initial issue that things get retried indefinitely with silent logs. If we're going to warn on every retry, we don't need to even count retries.

Instead, we might want to keep the `Debugf` and only start emitting `Warnf` here after `s.attempt` exceeds `maxRetries`. Otherwise, intermittent retries showing up as warnings in the logs might alarm users. We should probably only warn when it's a persistent problem. We already have debug logging if they want to see every retry.

That log change should go along with adding `log.Warnf`'s whenever we release a failed payload, like when we fail to send to the queue because the channel is full:
```
		select {
		case s.queue <- p:
			// ** Check s.attempt here **
			s.recordEvent(eventTypeRetry, stats)
			return
		default:
			// ** We should log here, including the err **
			// queue is full; since this is the oldest payload, we drop it
			s.releasePayload(p, eventTypeDropped, stats)
		}
```
And also in the default branch when it's an unknown error.
```
	default:
		// ** Same here **
		// this is a fatal error, we have to drop this payload
		s.releasePayload(p, eventTypeRejected, stats)
```

That way we are
* warning whenever we are retrying sends repeatedly
* warning whenever we drop payloads

Reply:
There is no need to add warnings in the places you mentioned since they already exist one level deeper - in `recordEvent` function that's invoked later in `releasePayload`